File: sdk/workbuddy-enterprises/rust-sdk/src/resources/usage.rs
use crate::client::{encode_path_segment, Client};
use crate::error::Result;
use crate::operations::validate_json_operation;
use crate::response::{parse_page, ApiResponse, Page};
use serde_json::{json, Value};

pub struct UsageResource<'a> {
    pub(crate) client: &'a Client,
}

impl UsageResource<'_> {
    pub fn get_quota_cycle(&self) -> Result<ApiResponse<Value>> {
        self.client.get_json("/openapi/usage/quota-cycle", &[])
    }

    pub fn get_default_quota(&self) -> Result<ApiResponse<Value>> {
        self.client.get_json("/openapi/usage/default-quota", &[])
    }

    pub fn update_default_quota(&self, body: Value) -> Result<ApiResponse<Value>> {
        validate_json_operation("usage-default-quota-update", &body)?;
        self.client
            .post_json("/openapi/usage/default-quota/update", &[], body)
    }

    pub fn query_members(&self, user_ids: &[&str]) -> Result<ApiResponse<Value>> {
        self.query_members_with_options(Some(user_ids), None, None, None, None, None)
    }

    #[allow(clippy::too_many_arguments)]
    pub fn query_members_with_options(
        &self,
        user_ids: Option<&[&str]>,
        user_names: Option<&[&str]>,
        page_num: Option<i64>,
        page_size: Option<i64>,
        start_time: Option<&str>,
        end_time: Option<&str>,
    ) -> Result<ApiResponse<Value>> {
        let mut body = member_selectors(user_ids, user_names, page_num, page_size);
        if let Some(start_time) = start_time {
            body.insert("startTime".into(), json!(start_time));
        }
        if let Some(end_time) = end_time {
            body.insert("endTime".into(), json!(end_time));
        }
        let body = Value::Object(body);
        validate_json_operation("usage-members-query", &body)?;
        self.client
            .post_json("/openapi/usage/members/query", &[], body)
    }

    pub fn query_member_limits(&self, user_ids: &[&str]) -> Result<ApiResponse<Value>> {
        self.query_member_limits_with_options(Some(user_ids), None, None, None)
    }

    pub fn query_member_limits_with_options(
        &self,
        user_ids: Option<&[&str]>,
        user_names: Option<&[&str]>,
        page_num: Option<i64>,
        page_size: Option<i64>,
    ) -> Result<ApiResponse<Value>> {
        let body = member_selectors(user_ids, user_names, page_num, page_size);
        let body = Value::Object(body);
        validate_json_operation("usage-members-limit-query", &body)?;
        self.client
            .post_json("/openapi/usage/members/limit-query", &[], body)
    }

    pub fn update_member_quota(&self, body: Value) -> Result<ApiResponse<Value>> {
        validate_json_operation("usage-members-quota-update", &body)?;
        self.client
            .post_json("/openapi/usage/members/quota/update", &[], body)
    }

    pub fn update_department_quota(
        &self,
        department_id: &str,
        body: Value,
    ) -> Result<ApiResponse<Value>> {
        validate_json_operation("usage-department-quota-update", &body)?;
        let department_id = encode_path_segment(department_id);
        self.client.post_json(
            &format!("/openapi/usage/departments/{department_id}/quota/update"),
            &[],
            body,
        )
    }

    pub fn query_member_details(&self, body: Value) -> Result<ApiResponse<Page<Value>>> {
        validate_json_operation("usage-members-detail", &body)?;
        let resp = self
            .client
            .post_json("/openapi/usage/members/detail", &[], body)?;
        let page = if resp.data.get("items").is_some() || resp.data.get("nextPageToken").is_some() {
            parse_page(resp.data)
        } else {
            Page {
                items: vec![],
                total_count: None,
                page: None,
                page_num: None,
                page_size: None,
                next_page_token: None,
                extra: resp.data,
            }
        };
        Ok(ApiResponse {
            data: page,
            code: resp.code,
            message: resp.message,
            request_id: resp.request_id,
            raw: resp.raw,
        })
    }
}

fn member_selectors(
    user_ids: Option<&[&str]>,
    user_names: Option<&[&str]>,
    page_num: Option<i64>,
    page_size: Option<i64>,
) -> serde_json::Map<String, Value> {
    let mut body = serde_json::Map::new();
    if let Some(user_ids) = user_ids {
        body.insert("userIds".into(), json!(user_ids));
    }
    if let Some(user_names) = user_names {
        body.insert("userNames".into(), json!(user_names));
    }
    if let Some(page_num) = page_num {
        body.insert("pageNum".into(), json!(page_num));
    }
    if let Some(page_size) = page_size {
        body.insert("pageSize".into(), json!(page_size));
    }
    body
}

File: sdk/workbuddy-enterprises/python-sdk/src/workbuddy_enterprise/resources/usage.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from workbuddy_enterprise._serialization import clean_dict
from workbuddy_enterprise.resources._base import Resource
from workbuddy_enterprise.response import ApiResponse, Page


class UsageResource(Resource):
    def get_quota_cycle(self) -> ApiResponse[dict[str, Any]]:
        return self._as_map(self._get("/openapi/usage/quota-cycle"))

    def get_default_quota(self) -> ApiResponse[dict[str, Any]]:
        return self._as_map(self._get("/openapi/usage/default-quota"))

    def update_default_quota(
        self,
        *,
        limit_type: str,
        new_limit: int | None = None,
        cycle_type: str | None = None,
        **fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        body = clean_dict(
            {
                "limitType": limit_type,
                "newLimit": new_limit,
                "cycleType": cycle_type,
                **fields,
            },
        )
        return self._as_map(self._post_json("/openapi/usage/default-quota/update", body=body))

    def query_members(
        self,
        *,
        user_ids: Sequence[str] | None = None,
        user_names: Sequence[str] | None = None,
        page_num: int | None = None,
        page_size: int | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
        **fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        body = clean_dict(
            {
                "userIds": list(user_ids) if user_ids is not None else None,
                "userNames": list(user_names) if user_names is not None else None,
                "pageNum": page_num,
                "pageSize": page_size,
                "startTime": start_time,
                "endTime": end_time,
                **fields,
            },
        )
        return self._as_map(self._post_json("/openapi/usage/members/query", body=body))

    def query_member_limits(
        self,
        *,
        user_ids: Sequence[str] | None = None,
        user_names: Sequence[str] | None = None,
        page_num: int | None = None,
        page_size: int | None = None,
        **fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        body = clean_dict(
            {
                "userIds": list(user_ids) if user_ids is not None else None,
                "userNames": list(user_names) if user_names is not None else None,
                "pageNum": page_num,
                "pageSize": page_size,
                **fields,
            },
        )
        return self._as_map(self._post_json("/openapi/usage/members/limit-query", body=body))

    def update_member_quota(
        self,
        *,
        limit_type: str,
        user_ids: Sequence[str] | None = None,
        user_names: Sequence[str] | None = None,
        new_limit: int | None = None,
        cycle_type: str | None = None,
        **fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        body = clean_dict(
            {
                "limitType": limit_type,
                "userIds": list(user_ids) if user_ids is not None else None,
                "userNames": list(user_names) if user_names is not None else None,
                "newLimit": new_limit,
                "cycleType": cycle_type,
                **fields,
            },
        )
        return self._as_map(self._post_json("/openapi/usage/members/quota/update", body=body))

    def update_department_quota(
        self,
        department_id: str,
        *,
        limit_type: str,
        new_limit: int | None = None,
        cycle_type: str | None = None,
        **fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        body = clean_dict(
            {
                "limitType": limit_type,
                "newLimit": new_limit,
                "cycleType": cycle_type,
                **fields,
            },
        )
        return self._as_map(
            self._post_json(
                f"/openapi/usage/departments/{self._segment(department_id)}/quota/update",
                body=body,
            ),
        )

    def query_member_details(
        self,
        *,
        time_range: Mapping[str, str],
        department_ids: Sequence[str] | None = None,
        user_ids: Sequence[str] | None = None,
        event_types: Sequence[str] | None = None,
        principal_types: Sequence[str] | None = None,
        page_num: int | None = None,
        page_size: int | None = None,
        group_id: str | None = None,
        version: int | None = None,
        page_token: str | None = None,
        **fields: Any,
    ) -> ApiResponse[Page[dict[str, Any]] | dict[str, Any]]:
        body = clean_dict(
            {
                "timeRange": dict(time_range),
                "departmentIds": list(department_ids) if department_ids is not None else None,
                "userIds": list(user_ids) if user_ids is not None else None,
                "eventTypes": list(event_types) if event_types is not None else None,
                "principalTypes": list(principal_types) if principal_types is not None else None,
                "pageNum": page_num,
                "pageSize": page_size,
                "groupId": group_id,
                "version": version,
                "pageToken": page_token,
                **fields,
            },
        )
        resp = self._post_json("/openapi/usage/members/detail", body=body)
        if isinstance(resp.data, dict) and ("items" in resp.data or "nextPageToken" in resp.data):
            data: Page[dict[str, Any]] | dict[str, Any] = self._as_page(resp).data
        else:
            data = self._as_map(resp).data
        return self._with_data(resp, data)
